File: examples/canonical-faker/python/profiles.py
from __future__ import annotations

import importlib.util
import sys
from dataclasses import replace
from pathlib import Path


def _load_domain_module():
    module_path = Path(__file__).resolve().parent.parent / "domain" / "generation_patterns.py"
    spec = importlib.util.spec_from_file_location(
        "canonical_faker_domain_generation_patterns", module_path
    )
    if spec is None or spec.loader is None:
        raise RuntimeError(f"unable to load domain module from {module_path}")
    module = importlib.util.module_from_spec(spec)
    sys.modules.setdefault(spec.name, module)
    spec.loader.exec_module(module)
    return module


_DOMAIN = _load_domain_module()
Profile = _DOMAIN.Profile


PROFILE_FACTORIES = {
    "smoke": Profile.SMOKE,
    "small": Profile.SMALL,
    "medium": Profile.MEDIUM,
    "large": Profile.LARGE,
}


def resolve_profile(name: str, seed: int | None = None) -> Profile:
    try:
        profile = PROFILE_FACTORIES[name]()
    except KeyError as exc:
        known = ", ".join(sorted(PROFILE_FACTORIES))
        raise ValueError(f"unknown profile {name!r}; expected one of: {known}") from exc
    if seed is not None:
        profile = replace(profile, seed=seed)
    return profile

File: examples/canonical-faker/ruby/lib/profiles.rb
class Profile
  attr_reader :name, :num_orgs, :num_users, :seed

  def initialize(name, num_orgs, num_users, seed)
    @name = name
    @num_orgs = num_orgs
    @num_users = num_users
    @seed = seed
  end

  SMOKE = Profile.new("smoke", 3, 10, 42).freeze
  SMALL = Profile.new("small", 20, 200, 1000).freeze
  MEDIUM = Profile.new("medium", 200, 5000, 5000).freeze
  LARGE = Profile.new("large", 2000, 50000, 9999).freeze

  def self.from_name(name)
    case name
    when "smoke" then SMOKE
    when "small" then SMALL
    when "medium" then MEDIUM
    when "large" then LARGE
    else SMOKE
    end
  end
end